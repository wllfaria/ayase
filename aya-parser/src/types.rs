#[derive(Debug, PartialEq, Eq, Clone, Copy)]
pub enum Operator {
    Add,
    Sub,
    Mul,
}

#[derive(Debug, PartialEq, Eq)]
pub enum Atom<'parser> {
    HexLiteral(&'parser str),
    Address(&'parser str),
    Register(&'parser str),
    Var(&'parser str),
    Operator(Operator),
    BinaryOp {
        lhs: Box<Atom<'parser>>,
        operator: Operator,
        rhs: Box<Atom<'parser>>,
    },
}

#[derive(Debug, PartialEq, Eq)]
pub enum Instruction<'parser> {
    MovLitReg(Atom<'parser>, Atom<'parser>),
    MovRegReg(Atom<'parser>, Atom<'parser>),
    MovRegMem(Atom<'parser>, Atom<'parser>),
    MovMemReg(Atom<'parser>, Atom<'parser>),
    MovLitMem(Atom<'parser>, Atom<'parser>),
    MovRegPtrReg(Atom<'parser>, Atom<'parser>),
    AddRegReg(Atom<'parser>, Atom<'parser>),
    AddLitReg(Atom<'parser>, Atom<'parser>),
    SubRegReg(Atom<'parser>, Atom<'parser>),
    SubLitReg(Atom<'parser>, Atom<'parser>),
    MulRegReg(Atom<'parser>, Atom<'parser>),
    MulLitReg(Atom<'parser>, Atom<'parser>),
    LshRegReg(Atom<'parser>, Atom<'parser>),
    LshLitReg(Atom<'parser>, Atom<'parser>),
    RshRegReg(Atom<'parser>, Atom<'parser>),
    RshLitReg(Atom<'parser>, Atom<'parser>),
    AndRegReg(Atom<'parser>, Atom<'parser>),
    AndLitReg(Atom<'parser>, Atom<'parser>),
    OrLitReg(Atom<'parser>, Atom<'parser>),
    OrRegReg(Atom<'parser>, Atom<'parser>),
    XorLitReg(Atom<'parser>, Atom<'parser>),
    XorRegReg(Atom<'parser>, Atom<'parser>),
    Inc(Atom<'parser>),
    Dec(Atom<'parser>),
    Not(Atom<'parser>),
    JeqLit(Atom<'parser>, Atom<'parser>),
    JeqReg(Atom<'parser>, Atom<'parser>),
    JgtLit(Atom<'parser>, Atom<'parser>),
    JgtReg(Atom<'parser>, Atom<'parser>),
    JneLit(Atom<'parser>, Atom<'parser>),
    JneReg(Atom<'parser>, Atom<'parser>),
    JgeLit(Atom<'parser>, Atom<'parser>),
    JgeReg(Atom<'parser>, Atom<'parser>),
    JleLit(Atom<'parser>, Atom<'parser>),
    JleReg(Atom<'parser>, Atom<'parser>),
    JltLit(Atom<'parser>, Atom<'parser>),
    JltReg(Atom<'parser>, Atom<'parser>),
    PshLit(Atom<'parser>),
    PshReg(Atom<'parser>),
    Pop(Atom<'parser>),
    CalLit(Atom<'parser>),
    CalReg(Atom<'parser>),
    Ret,
    Hlt,
}
