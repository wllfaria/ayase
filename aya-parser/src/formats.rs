use nom::branch::alt;
use nom::bytes::complete::tag_no_case;
use nom::character::complete::{char, space0, space1};
use nom::sequence::preceded;
use nom::IResult;

use crate::common::{address, hex_literal, register};
use crate::expressions::square_bracketed_expr;
use crate::types::{Atom, Instruction};

pub fn lit_reg<'parser, M: Fn(Atom<'parser>, Atom<'parser>) -> Instruction<'parser>>(
    input: &'parser str,
    prefix: &'parser str,
    mapper: M,
) -> IResult<&'parser str, Instruction<'parser>> {
    let (input, _) = tag_no_case(prefix)(input)?;
    let (input, _) = space1(input)?;

    let (input, lhs) = register(input)?;

    let (input, _) = space0(input)?;
    let (input, _) = char(',')(input)?;
    let (input, _) = space0(input)?;

    let (input, rhs) = alt((hex_literal, square_bracketed_expr))(input)?;
    let (input, _) = space0(input)?;

    Ok((input, mapper(lhs, rhs)))
}

pub fn reg_reg<'parser, M: Fn(Atom<'parser>, Atom<'parser>) -> Instruction<'parser>>(
    input: &'parser str,
    prefix: &'parser str,
    mapper: M,
) -> IResult<&'parser str, Instruction<'parser>> {
    let (input, _) = tag_no_case(prefix)(input)?;
    let (input, _) = space1(input)?;

    let (input, lhs) = register(input)?;

    let (input, _) = space0(input)?;
    let (input, _) = char(',')(input)?;
    let (input, _) = space0(input)?;

    let (input, rhs) = register(input)?;
    let (input, _) = space0(input)?;

    Ok((input, mapper(lhs, rhs)))
}

pub fn reg_mem<'parser, M: Fn(Atom<'parser>, Atom<'parser>) -> Instruction<'parser>>(
    input: &'parser str,
    prefix: &'parser str,
    mapper: M,
) -> IResult<&'parser str, Instruction<'parser>> {
    let (input, _) = tag_no_case(prefix)(input)?;
    let (input, _) = space1(input)?;

    let (input, lhs) = alt((address, preceded(char('&'), square_bracketed_expr)))(input)?;

    let (input, _) = space0(input)?;
    let (input, _) = char(',')(input)?;
    let (input, _) = space0(input)?;

    let (input, rhs) = register(input)?;
    let (input, _) = space0(input)?;

    Ok((input, mapper(lhs, rhs)))
}

pub fn mem_reg<'parser, M: Fn(Atom<'parser>, Atom<'parser>) -> Instruction<'parser>>(
    input: &'parser str,
    prefix: &'parser str,
    mapper: M,
) -> IResult<&'parser str, Instruction<'parser>> {
    let (input, _) = tag_no_case(prefix)(input)?;
    let (input, _) = space1(input)?;

    let (input, lhs) = register(input)?;

    let (input, _) = space0(input)?;
    let (input, _) = char(',')(input)?;
    let (input, _) = space0(input)?;

    let (input, rhs) = alt((address, preceded(char('&'), square_bracketed_expr)))(input)?;
    let (input, _) = space0(input)?;

    Ok((input, mapper(lhs, rhs)))
}

pub fn lit_mem<'parser, M: Fn(Atom<'parser>, Atom<'parser>) -> Instruction<'parser>>(
    input: &'parser str,
    prefix: &'parser str,
    mapper: M,
) -> IResult<&'parser str, Instruction<'parser>> {
    let (input, _) = tag_no_case(prefix)(input)?;
    let (input, _) = space1(input)?;

    let (input, lhs) = alt((address, preceded(char('&'), square_bracketed_expr)))(input)?;

    let (input, _) = space0(input)?;
    let (input, _) = char(',')(input)?;
    let (input, _) = space0(input)?;

    let (input, rhs) = alt((hex_literal, square_bracketed_expr))(input)?;
    let (input, _) = space0(input)?;

    Ok((input, mapper(lhs, rhs)))
}

pub fn reg_ptr_reg<'parser, M: Fn(Atom<'parser>, Atom<'parser>) -> Instruction<'parser>>(
    input: &'parser str,
    prefix: &'parser str,
    mapper: M,
) -> IResult<&'parser str, Instruction<'parser>> {
    let (input, _) = tag_no_case(prefix)(input)?;
    let (input, _) = space1(input)?;

    let (input, lhs) = register(input)?;

    let (input, _) = space0(input)?;
    let (input, _) = char(',')(input)?;
    let (input, _) = space0(input)?;

    let (input, rhs) = preceded(char('&'), register)(input)?;
    let (input, _) = space0(input)?;

    Ok((input, mapper(lhs, rhs)))
}

pub fn single_reg<'parser, M: Fn(Atom<'parser>) -> Instruction<'parser>>(
    input: &'parser str,
    prefix: &'parser str,
    mapper: M,
) -> IResult<&'parser str, Instruction<'parser>> {
    let (input, _) = tag_no_case(prefix)(input)?;
    let (input, _) = space1(input)?;

    let (input, reg) = register(input)?;
    let (input, _) = space0(input)?;

    Ok((input, mapper(reg)))
}

pub fn single_lit<'parser, M: Fn(Atom<'parser>) -> Instruction<'parser>>(
    input: &'parser str,
    prefix: &'parser str,
    mapper: M,
) -> IResult<&'parser str, Instruction<'parser>> {
    let (input, _) = tag_no_case(prefix)(input)?;
    let (input, _) = space1(input)?;

    let (input, lit) = alt((hex_literal, square_bracketed_expr))(input)?;
    let (input, _) = space0(input)?;

    Ok((input, mapper(lit)))
}

pub fn no_arg<'parser, M: Fn() -> Instruction<'parser>>(
    input: &'parser str,
    prefix: &'parser str,
    mapper: M,
) -> IResult<&'parser str, Instruction<'parser>> {
    let (input, _) = tag_no_case(prefix)(input)?;
    let (input, _) = space0(input)?;

    Ok((input, mapper()))
}
