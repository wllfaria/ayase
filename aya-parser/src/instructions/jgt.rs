use nom::branch::alt;
use nom::IResult;

use crate::formats::{lit_mem, reg_mem};
use crate::types::Instruction;

pub fn parse_jgt(input: &str) -> IResult<&str, Instruction> {
    alt((
        |input| lit_mem(input, "jgt", Instruction::JgtLit),
        |input| reg_mem(input, "jgt", Instruction::JgtReg),
    ))(input)
}

#[cfg(test)]
mod tests {
    use super::*;

    fn run_instruction(input: &str) -> (&str, Instruction) {
        parse_jgt(input).unwrap()
    }

    #[test]
    fn test_jgt_reg() {
        let input = "jgt &c0d3, r2";
        let (input, instruction) = run_instruction(input);
        insta::assert_debug_snapshot!(instruction);
        assert!(input.is_empty());
    }

    #[test]
    fn test_jgt_lit() {
        let input = "jgt &c0d3, $0303";
        let (input, instruction) = run_instruction(input);
        insta::assert_debug_snapshot!(instruction);
        assert!(input.is_empty());
    }
}
