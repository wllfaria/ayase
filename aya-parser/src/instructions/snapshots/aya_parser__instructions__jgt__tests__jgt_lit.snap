---
source: aya-parser/src/instructions/jgt.rs
expression: instruction
---
JgtLit(
    Address(
        "c0d3",
    ),
    HexLiteral(
        "0303",
    ),
)
